File: packages/ollama-client/src/gemini.ts
/** Shared LLM types and Gemini client for agents (Node). */

export type AiProviderId = "local" | "gemini";

/** Default / medium effort — agentic loops, speed + intelligence. */
export const GEMINI_FLASH = "gemini-3.6-flash";
/** High+ effort — strongest Gemini for complex agent planning/critique. */
export const GEMINI_PRO = "gemini-3.1-pro-preview";

export interface ChatMessage {
  role: "system" | "user" | "assistant";
  content: string;
}

export interface ChatOptions {
  model: string;
  temperature?: number;
  numCtx?: number;
  format?: "json";
  /** Abort chat if the provider does not respond in time (default 90s). */
  timeoutMs?: number;
}

export interface LlmClient {
  chat(messages: ChatMessage[], options: ChatOptions): Promise<string>;
  isAvailable(): Promise<boolean>;
  listModels(): Promise<string[]>;
  pullModel(model: string, onProgress?: (status: string) => void): Promise<void>;
}

export function geminiModelsForEffort(effort: string): {
  plannerModel: string;
  extractorModel: string;
  criticModel?: string;
} {
  const heavy =
    effort === "high" ||
    effort === "super_high" ||
    effort === "ultra_high" ||
    effort === "pro" ||
    effort === "max";
  // Quality-first: Pro for plan + extract + critic on heavy runs; Flash only for light extract.
  return {
    plannerModel: GEMINI_PRO,
    extractorModel: heavy ? GEMINI_PRO : GEMINI_FLASH,
    criticModel: heavy ? GEMINI_PRO : undefined,
  };
}

/** Per-call timeout: larger models need longer; Flash/Ollama small = 3 min. */
export function defaultLlmTimeoutMs(model: string): number {
  const m = model.toLowerCase();
  if (m.includes("gemini") && m.includes("pro")) return 360_000;
  if (m.includes("gemini")) return 240_000;
  if (/\b(70b|72b|32b)\b/.test(m)) return 600_000;
  if (/\b(14b|13b|15b)\b/.test(m)) return 420_000;
  if (/\b(7b|8b|9b)\b/.test(m)) return 300_000;
  return 180_000;
}

function buildGeminiBody(
  messages: ChatMessage[],
  temperature: number | undefined,
  responseJson: boolean
): Record<string, unknown> {
  const systemParts: string[] = [];
  const contents: { role: string; parts: { text: string }[] }[] = [];

  for (const msg of messages) {
    if (msg.role === "system") {
      if (msg.content.trim()) systemParts.push(msg.content);
      continue;
    }
    const role = msg.role === "assistant" ? "model" : "user";
    const last = contents[contents.length - 1];
    if (last && last.role === role) {
      last.parts.push({ text: msg.content });
    } else {
      contents.push({ role, parts: [{ text: msg.content }] });
    }
  }

  const body: Record<string, unknown> = {
    contents,
    generationConfig: {
      temperature: temperature ?? 0.5,
    },
  };
  if (systemParts.length) {
    body.systemInstruction = { parts: [{ text: systemParts.join("\n\n") }] };
  }
  if (responseJson) {
    (body.generationConfig as Record<string, unknown>).responseMimeType = "application/json";
  }
  return body;
}

function mapGeminiError(status: number, body: string): string {
  const lower = body.toLowerCase();
  if (status === 400 && (lower.includes("api key") || lower.includes("api_key"))) {
    return "Gemini API key inválida. Revisa la clave en Ajustes.";
  }
  if (status === 401 || status === 403) {
    return "Gemini rechazó la API key (no autorizada). Revisa la clave en Ajustes.";
  }
  if (status === 429) {
    return "Cuota de Gemini agotada o demasiadas peticiones. Espera un momento o revisa tu plan.";
  }
  if (status >= 500) {
    return `Gemini no está disponible temporalmente (HTTP ${status}).`;
  }
  const truncated = body.slice(0, 280);
  return truncated
    ? `Gemini rechazó la petición: ${truncated}`
    : `Gemini rechazó la petición (HTTP ${status})`;
}

export class GeminiClient implements LlmClient {
  private apiKey: string;
  private baseUrl = "https://generativelanguage.googleapis.com/v1beta";

  constructor(apiKey: string) {
    this.apiKey = apiKey.trim();
  }

  async isAvailable(): Promise<boolean> {
    return Boolean(this.apiKey);
  }

  async listModels(): Promise<string[]> {
    return [GEMINI_FLASH, GEMINI_PRO];
  }

  async pullModel(_model: string, _onProgress?: (status: string) => void): Promise<void> {
    // Cloud models — nothing to pull.
  }

  async chat(messages: ChatMessage[], options: ChatOptions): Promise<string> {
    const model = options.model.startsWith("gemini") ? options.model : GEMINI_FLASH;
    const timeoutMs = options.timeoutMs ?? defaultLlmTimeoutMs(model);
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), timeoutMs);
    const body = buildGeminiBody(messages, options.temperature, options.format === "json");
    const url = `${this.baseUrl}/models/${model}:generateContent`;

    try {
      let lastErr = "";
      for (let attempt = 0; attempt < 3; attempt++) {
        const res = await fetch(url, {
          method: "POST",
          headers: {
            "Content-Type": "application/json",
            "x-goog-api-key": this.apiKey,
          },
          signal: controller.signal,
          body: JSON.stringify(body),
        });
        const text = await res.text();
        if (res.status === 429 || res.status >= 500) {
          lastErr = mapGeminiError(res.status, text);
          if (attempt < 2) {
            await new Promise((r) => setTimeout(r, 400 * (attempt + 1) ** 2));
            continue;
          }
          throw new Error(lastErr);
        }
        if (!res.ok) {
          throw new Error(mapGeminiError(res.status, text));
        }
        const data = JSON.parse(text) as {
          candidates?: { content?: { parts?: { text?: string }[] } }[];
        };
        const out =
          data.candidates?.[0]?.content?.parts?.map((p) => p.text ?? "").join("") ?? "";
        if (!out) throw new Error("Gemini devolvió una respuesta vacía");
        return out;
      }
      throw new Error(lastErr || "Gemini falló");
    } catch (err) {
      if (err instanceof Error && err.name === "AbortError") {
        throw new Error(`Gemini chat timed out after ${timeoutMs}ms (${model})`);
      }
      throw err;
    } finally {
      clearTimeout(timer);
    }
  }
}

export type CreateLlmClientOptions = {
  provider?: AiProviderId | string;
  apiKey?: string;
  /** Existing Ollama-compatible client for local mode. */
  localClient?: LlmClient;
};

export function createLlmClient(options: CreateLlmClientOptions = {}): LlmClient {
  const provider = (options.provider ?? "local").toLowerCase();
  if (provider === "gemini") {
    const key = options.apiKey?.trim();
    if (!key) {
      throw new Error("Gemini seleccionado pero no hay API key. Configúrala en Ajustes.");
    }
    return new GeminiClient(key);
  }
  if (options.localClient) return options.localClient;
  throw new Error("Local LLM client required (pass localClient)");
}

export function createLlmClientFromEnv(localClient: LlmClient): LlmClient {
  const env =
    typeof process !== "undefined" && process.env ? process.env : ({} as NodeJS.ProcessEnv);
  const provider = (env.AIIA_LLM_PROVIDER ?? "local").toLowerCase();
  const apiKey = env.AIIA_GEMINI_API_KEY;
  if (provider === "gemini") {
    return createLlmClient({ provider: "gemini", apiKey });
  }
  return localClient;
}

File: apps/desktop/src-tauri/src/gemini.rs
//! Google Gemini API client (user-provided API key). Streaming + sync chat.

use std::time::Duration;

use serde::Serialize;
use tauri::{AppHandle, Emitter};

use crate::ollama::{is_stream_cancelled, ChatStreamEvent};

pub const GEMINI_SITE_ID: &str = "aiia.gemini";
pub const AI_PROVIDER_SETTING: &str = "ai_provider";
pub const DEFAULT_GEMINI_FLASH: &str = "gemini-3.6-flash";
pub const DEFAULT_GEMINI_PRO: &str = "gemini-3.1-pro-preview";

const GEMINI_API: &str = "https://generativelanguage.googleapis.com/v1beta";

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum AiProvider {
    Local,
    Gemini,
}

impl AiProvider {
    pub fn parse(value: Option<&str>) -> Self {
        match value.map(|s| s.trim().to_ascii_lowercase()).as_deref() {
            Some("gemini") => Self::Gemini,
            _ => Self::Local,
        }
    }

    pub fn as_str(self) -> &'static str {
        match self {
            Self::Local => "local",
            Self::Gemini => "gemini",
        }
    }
}

pub fn gemini_model_for_mode(mode: &str) -> &'static str {
    match mode {
        "pro" | "max" | "high" | "super_high" | "ultra_high" => DEFAULT_GEMINI_PRO,
        _ => DEFAULT_GEMINI_FLASH,
    }
}

fn map_gemini_http_error(status: u16, body: &str) -> String {
    let lower = body.to_ascii_lowercase();
    if status == 400 && (lower.contains("api key") || lower.contains("api_key")) {
        return "Gemini API key inválida. Revisa la clave en Ajustes.".to_string();
    }
    if status == 401 || status == 403 {
        return "Gemini rechazó la API key (no autorizada). Revisa la clave en Ajustes.".to_string();
    }
    if status == 429 {
        return "Cuota de Gemini agotada o demasiadas peticiones. Espera un momento o revisa tu plan en Google AI Studio.".to_string();
    }
    if status >= 500 {
        return format!("Gemini no está disponible temporalmente (HTTP {status}). Inténtalo de nuevo.");
    }
    if body.is_empty() {
        format!("Gemini rechazó la petición (HTTP {status})")
    } else {
        let truncated: String = body.chars().take(280).collect();
        format!("Gemini rechazó la petición: {truncated}")
    }
}

/// Convert Ollama-style messages to Gemini `systemInstruction` + `contents`.
pub fn build_gemini_body(
    messages: &[serde_json::Value],
    temperature: Option<f64>,
    response_json: bool,
) -> Result<serde_json::Value, String> {
    let mut system_parts: Vec<String> = Vec::new();
    let mut contents: Vec<serde_json::Value> = Vec::new();

    for msg in messages {
        let role = msg
            .get("role")
            .and_then(|r| r.as_str())
            .unwrap_or("user");
        let content = msg
            .get("content")
            .and_then(|c| c.as_str())
            .unwrap_or("")
            .to_string();
        let images = msg
            .get("images")
            .and_then(|v| v.as_array())
            .cloned()
            .unwrap_or_default();

        if role == "system" {
            if !content.trim().is_empty() {
                system_parts.push(content);
            }
            continue;
        }

        let gemini_role = if role == "assistant" { "model" } else { "user" };
        let mut parts: Vec<serde_json::Value> = Vec::new();
        if !content.is_empty() {
            parts.push(serde_json::json!({ "text": content }));
        }
        for img in images {
            let Some(b64) = img.as_str() else { continue };
            let (mime, data) = split_data_url(b64);
            parts.push(serde_json::json!({
                "inlineData": {
                    "mimeType": mime,
                    "data": data,
                }
            }));
        }
        if parts.is_empty() {
            continue;
        }

        // Merge consecutive same-role turns (Gemini requirement).
        if let Some(last) = contents.last_mut() {
            if last.get("role").and_then(|r| r.as_str()) == Some(gemini_role) {
                if let Some(arr) = last.get_mut("parts").and_then(|p| p.as_array_mut()) {
                    arr.extend(parts);
                    continue;
                }
            }
        }

        contents.push(serde_json::json!({
            "role": gemini_role,
            "parts": parts,
        }));
    }

    if contents.is_empty() {
        return Err("No hay mensajes para enviar a Gemini".to_string());
    }

    let mut body = serde_json::json!({
        "contents": contents,
        "generationConfig": {
            "temperature": temperature.unwrap_or(0.7),
        }
    });

    if !system_parts.is_empty() {
        body["systemInstruction"] = serde_json::json!({
            "parts": [{ "text": system_parts.join("\n\n") }]
        });
    }

    if response_json {
        body["generationConfig"]["responseMimeType"] = serde_json::json!("application/json");
    }

    Ok(body)
}

fn split_data_url(raw: &str) -> (&str, &str) {
    if let Some(rest) = raw.strip_prefix("data:") {
        if let Some((meta, data)) = rest.split_once(",") {
            let mime = meta.split(';').next().unwrap_or("image/jpeg");
            return (mime, data);
        }
    }
    ("image/jpeg", raw)
}

fn extract_text_delta(value: &serde_json::Value) -> String {
    value
        .get("candidates")
        .and_then(|c| c.as_array())
        .and_then(|arr| arr.first())
        .and_then(|cand| cand.get("content"))
        .and_then(|c| c.get("parts"))
        .and_then(|p| p.as_array())
        .map(|parts| {
            parts
                .iter()
                .filter_map(|part| part.get("text").and_then(|t| t.as_str()))
                .collect::<Vec<_>>()
                .join("")
        })
        .unwrap_or_default()
}

async fn sleep_backoff(attempt: u32) {
    let ms = match attempt {
        0 => 400,
        1 => 1200,
        _ => 3000,
    };
    tokio::time::sleep(Duration::from_millis(ms)).await;
}

pub async fn gemini_chat(
    api_key: &str,
    model: String,
    messages: Vec<serde_json::Value>,
    temperature: Option<f64>,
    format: Option<String>,
) -> Result<String, String> {
    let response_json = format.as_deref() == Some("json");
    let body = build_gemini_body(&messages, temperature, response_json)?;
    let url = format!("{GEMINI_API}/models/{model}:generateContent");

    let client = reqwest::Client::builder()
        .timeout(Duration::from_secs(if model.to_lowercase().contains("pro") {
            300
        } else {
            180
        }))
        .build()
        .map_err(|e| e.to_string())?;

    let mut last_err = String::new();
    for attempt in 0..3u32 {
        let res = client
            .post(&url)
            .header("x-goog-api-key", api_key)
            .header("Content-Type", "application/json")
            .json(&body)
            .send()
            .await
            .map_err(|e| format!("No se pudo conectar con Gemini: {e}"))?;

        let status = res.status().as_u16();
        let text = res.text().await.unwrap_or_default();
        if status == 429 || status >= 500 {
            last_err = map_gemini_http_error(status, &text);
            if attempt < 2 {
                sleep_backoff(attempt).await;
                continue;
            }
            return Err(last_err);
        }
        if !(200..300).contains(&status) {
            return Err(map_gemini_http_error(status, &text));
        }

        let data: serde_json::Value =
            serde_json::from_str(&text).map_err(|e| format!("Respuesta Gemini inválida: {e}"))?;
        let out = extract_text_delta(&data);
        if out.is_empty() {
            return Err("Gemini devolvió una respuesta vacía".to_string());
        }
        return Ok(out);
    }
    Err(last_err)
}

pub async fn gemini_chat_stream(
    app: AppHandle,
    api_key: String,
    stream_id: String,
    model: String,
    messages: Vec<serde_json::Value>,
    temperature: Option<f64>,
    cancel_set: std::sync::Arc<std::sync::Mutex<std::collections::HashSet<String>>>,
) -> Result<(), String> {
    let emit_done = |delta: String, cancelled: bool, error: Option<String>| {
        let _ = app.emit(
            "chat-stream",
            ChatStreamEvent {
                stream_id: stream_id.clone(),
                delta,
                done: true,
                cancelled,
                error,
            },
        );
    };

    let body = match build_gemini_body(&messages, temperature, false) {
        Ok(b) => b,
        Err(e) => {
            emit_done(String::new(), false, Some(e.clone()));
            return Err(e);
        }
    };

    let url = format!("{GEMINI_API}/models/{model}:streamGenerateContent?alt=sse");
    let client = reqwest::Client::builder()
        .timeout(Duration::from_secs(900))
        .build()
        .map_err(|e| e.to_string())?;

    let mut last_err = String::new();
    let mut res_opt = None;
    for attempt in 0..3u32 {
        if is_stream_cancelled(&cancel_set, &stream_id) {
            let _ = cancel_set.lock().map(|mut set| set.remove(&stream_id));
            emit_done(String::new(), true, None);
            return Ok(());
        }

        let res = match client
            .post(&url)
            .header("x-goog-api-key", &api_key)
            .header("Content-Type", "application/json")
            .json(&body)
            .send()
            .await
        {
            Ok(r) => r,
            Err(e) => {
                last_err = format!("No se pudo conectar con Gemini: {e}");
                if attempt < 2 {
                    sleep_backoff(attempt).await;
                    continue;
                }
                emit_done(String::new(), false, Some(last_err.clone()));
                return Err(last_err);
            }
        };

        let status = res.status().as_u16();
        if status == 429 || status >= 500 {
            let text = res.text().await.unwrap_or_default();
            last_err = map_gemini_http_error(status, &text);
            if attempt < 2 {
                sleep_backoff(attempt).await;
                continue;
            }
            emit_done(String::new(), false, Some(last_err.clone()));
            return Err(last_err);
        }
        if !(200..300).contains(&status) {
            let text = res.text().await.unwrap_or_default();
            let err = map_gemini_http_error(status, &text);
            emit_done(String::new(), false, Some(err.clone()));
            return Err(err);
        }
        res_opt = Some(res);
        break;
    }

    let Some(res) = res_opt else {
        emit_done(String::new(), false, Some(last_err.clone()));
        return Err(last_err);
    };

    use futures_util::StreamExt;
    let mut stream = res.bytes_stream();
    let mut buffer = Vec::new();

    while let Some(chunk) = stream.next().await {
        if is_stream_cancelled(&cancel_set, &stream_id) {
            let _ = cancel_set.lock().map(|mut set| set.remove(&stream_id));
            emit_done(String::new(), true, None);
            return Ok(());
        }

        let chunk = chunk.map_err(|e| {
            let err = format!("Error de red Gemini: {e}");
            emit_done(String::new(), false, Some(err.clone()));
            err
        })?;
        buffer.extend_from_slice(&chunk);

        while let Some(pos) = buffer.iter().position(|&b| b == b'\n') {
            let line_bytes: Vec<u8> = buffer.drain(..=pos).collect();
            let line = String::from_utf8_lossy(&line_bytes).trim().to_string();
            if line.is_empty() || line.starts_with(':') {
                continue;
            }
            let payload = if let Some(rest) = line.strip_prefix("data:") {
                rest.trim()
            } else {
                continue;
            };
            if payload.is_empty() || payload == "[DONE]" {
                continue;
            }
            let Ok(parsed) = serde_json::from_str::<serde_json::Value>(payload) else {
                continue;
            };
            let delta = extract_text_delta(&parsed);
            if !delta.is_empty() {
                let _ = app.emit(
                    "chat-stream",
                    ChatStreamEvent {
                        stream_id: stream_id.clone(),
                        delta,
                        done: false,
                        cancelled: false,
                        error: None,
                    },
                );
            }
        }
    }

    emit_done(String::new(), false, None);
    Ok(())
}

pub async fn test_gemini_api_key(api_key: &str) -> Result<(), String> {
    let messages = vec![serde_json::json!({
        "role": "user",
        "content": "Reply with exactly: ok"
    })];
    let out = gemini_chat(
        api_key,
        DEFAULT_GEMINI_FLASH.to_string(),
        messages,
        Some(0.0),
        None,
    )
    .await?;
    if out.trim().is_empty() {
        return Err("Gemini respondió vacío".to_string());
    }
    Ok(())
}

#[derive(Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct AiProviderStatus {
    pub provider: String,
    pub has_gemini_key: bool,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn maps_system_and_roles() {
        let messages = vec![
            serde_json::json!({"role": "system", "content": "Be brief"}),
            serde_json::json!({"role": "user", "content": "Hi"}),
            serde_json::json!({"role": "assistant", "content": "Hello"}),
            serde_json::json!({"role": "user", "content": "Again"}),
        ];
        let body = build_gemini_body(&messages, Some(0.2), false).unwrap();
        assert!(body.get("systemInstruction").is_some());
        let contents = body.get("contents").unwrap().as_array().unwrap();
        assert_eq!(contents.len(), 3);
        assert_eq!(contents[0]["role"], "user");
        assert_eq!(contents[1]["role"], "model");
        assert_eq!(contents[2]["role"], "user");
    }

    #[test]
    fn merges_consecutive_user_turns() {
        let messages = vec![
            serde_json::json!({"role": "user", "content": "A"}),
            serde_json::json!({"role": "user", "content": "B"}),
        ];
        let body = build_gemini_body(&messages, None, false).unwrap();
        let contents = body.get("contents").unwrap().as_array().unwrap();
        assert_eq!(contents.len(), 1);
        assert_eq!(contents[0]["parts"].as_array().unwrap().len(), 2);
    }

    #[test]
    fn maps_images_as_inline_data_camel_case() {
        let messages = vec![serde_json::json!({
            "role": "user",
            "content": "What is this?",
            "images": ["data:image/png;base64,aaa"]
        })];
        let body = build_gemini_body(&messages, None, false).unwrap();
        let parts = body["contents"][0]["parts"].as_array().unwrap();
        assert!(parts.iter().any(|p| p.get("inlineData").is_some()));
        assert!(parts.iter().any(|p| {
            p.get("inlineData")
                .and_then(|d| d.get("mimeType"))
                .and_then(|m| m.as_str())
                == Some("image/png")
        }));
        assert!(parts.iter().all(|p| p.get("inline_data").is_none()));
    }
}
